composer/trainer/trainer.py

```diff
@@ -140,6 +246,140 @@ def _unpack_evaluators(
     return evaluators
 
 
+def _unpack_grad_accum(grad_accum: Union[int, str], device: Device) -> Tuple[int, bool]:
+    # Set initial grad_accum to 1 if using adaptive
+    adaptive_grad_accum = grad_accum == "auto"
+    if adaptive_grad_accum:
+        grad_accum = 1
+        warnings.warn(textwrap.dedent("""Setting `grad_accum='auto'` is an experimental feature
+            which may cause uncaught Cuda Out of Memory errors. In this case, please manually
+            set grad_accum explicitly to an integer instead.
+            """),
+                      category=UserWarning)
+    # Cannot use adaptive grad accum on CPU
+    if isinstance(device, DeviceCPU) and adaptive_grad_accum:
+        raise ValueError("Cannot use adaptive grad_accum on CPU. Please set grad_accum >= 1")
+    # grad_accum should be int as we've already handled "auto" case
+    if isinstance(grad_accum, str):
+        raise ValueError("grad_accum must be an int or ``auto``")
+    return grad_accum, adaptive_grad_accum
+
+
+def _is_cuda_oom(e: RuntimeError):
+    """Determines if error is CUDA Out of Memory and if adaptive_grad_accum is enabled."""
+    return "CUDA out of memory" in str(e)
+
+
+def _unpack_device(device: Optional[Union[str, Device]], deepspeed_enabled: bool):
```

Comment:
`deepspeed_enabled` is not used here?

rename to `_get_device()`

Reply:
Done